module/zfs/arc.c

```diff
@@ -9290,6 +9354,49 @@ l2arc_stop(void)
 	mutex_exit(&l2arc_feed_thr_lock);
 }
 
+/*
+ * Punches out TRIM threads for the L2ARC devices in a spa and assigns them
+ * to vd->vdev_trim_thread variable. This facilitates the management of
+ * trimming the whole cache device using TRIM_TYPE_MANUAL upon addition
+ * to a pool or pool creation or when the header of the device is invalid.
+ */
+void
+l2arc_spa_trim(spa_t *spa)
+{
+	ASSERT(MUTEX_HELD(&spa_namespace_lock));
+
+	/*
+	 * Locate the spa's l2arc devices and kick off TRIM threads.
+	 */
+	for (int i = 0; i < spa->spa_l2cache.sav_count; i++) {
+		l2arc_dev_t *dev =
+		    l2arc_vdev_get(spa->spa_l2cache.sav_vdevs[i]);
+		if (dev == NULL) {
+			/* Don't attempt TRIM if the vdev is UNAVAIL */
+			continue;
+		}
+
+		if (dev->l2ad_trim_all) {
+			dev->l2ad_vdev->vdev_trim_thread = thread_create(NULL,
```

Comment:
Using the `vd->vdev_trim_*` members for this makes sense, but it does mean that a normal manual trim cannot be running when `vdev_trim_simple(, TRIM_TYPE_MANUAL)` is called on the same vdev.  They're mutually exclusive operations.  Given that, I think it would be best to move this threading logic entirely in to `vdev_trim.c` and model it after the manual trim threads.  Specifically adding a new `vdev_trim_l2arc()`  function which kicks off a `vdev_trim_l2arc_thread()` that does what `l2arc_dev_trim` and `vdev_trim_simple` current do.

Then `vdev_trim_simple()` could be restored to a more generic implementation which doesn't modify any of the `vd->vdev_trim_*` state.